### ISSUE TITLE
Rename Local

### ISSUE THREAD

**Issue by stevemk14ebr** [dnSpyEx/dnSpy#99]
Is it possible to allow the renaming of local, without first opening the C# source editor and re-compiling. Even if this is just per session and not saved back to the module (via the mdwriter), this feature would be a huge help in analysis of complex .NET functions.

**Comment by GazziFX** [comment]
Isn't that what you're looking for?
![image](https://user-images.githubusercontent.com/43498426/178055997-b9fc0428-8592-4c92-9d76-fd1802e9fdb9.png)

**Comment by ElektroKill** [comment]
Hello, there are two obstacles when it to comes to local renaming. One of them is the place where the data is stored and the second one is that certain variables in the decompilation are decompiler generated and thus metadata for them does not exist.

The current ways of renaming locals involve recompiling the decompiled code as you mentioned in the initial post or editing the name of the local in the edit method body screen. It is worth pointing out that locals in the edit body screen are the ones present in metadata and not necessarily a direct representation of the locals in the decompiled code so the proposed solution by @GazziFX might not always work. The decompiler can create its own locals when certain code cannot be expressed nicely without one or it can inline locals which exist in metadata in such a way they are no longer visible in the disassembly. The other way of recompiling decompiled code is time confusing and does not often preserve the semantics of the code as the decompiled code produced by the ILSpy 2.4 engine is far from perfect.

The solution proposed by @GazziFX works well if we are not dealing with decompiler-generated locals or inlined locals. It even allows for preserving the names across multiple sessions if a PDB file is present next to the original file as then the local names can be written into the PDB file and read from it in a new session. dnSpy currently does not currently allow the user to generate a fresh PDB file and is only capable of modifying existing ones.

The best solution for this would be to implement a "project" system, similar to what Ghidra and IDA do. This would enable the creation of a project file for a specific target being reversed which would store information regarding new symbol names and maybe even comments and other annotations if we want to go all out. This would also allow for easy reversing engineering without having to modify the original file in any way. This would also allow us to map any local variables present in the decompilation to be renamed and the new names to be stored cross-session in a separate file. This is something that has been on my to-do list for a while but designing a project format that allows for the same level of extensibility when it comes to extensions as all the other components of dnSpy is challenging.

If we are talking about per-session renaming, this is also not easy to implement as there is currently no way to tell the decompiler what names to use for the artificial locals. Another issue is that the UI layer only has access to the `SourceLocal` objects which are generated per decompilation run and are immutable.

In conclusion, currently, the best way of renaming locals without dealing with the side effects of recompiling decompiled code is through the usage of the edit method body feature. If no PDB file is present the local names are per session. If a PDB is present the local names are written to the PDB file when saving the module.

Related issue: #11

Feedback, suggestions, or questions are welcome as always :)
